src/main/java/world/data/jdbc/internal/transport/FileBackedInputStream.java

```diff
@@ -0,0 +1,199 @@
+ * for the entire download to complete before they can make progress--a semaphore is used to ensure readers don't
+ * get ahead of the download thread.</li>
+ * </ol>
+ */
+class FileBackedInputStream extends InputStream {
+    private InputStream memIn;
+    private final Sync sync;
+    private File file;
+    private final InputStream fileIn;
+    /** Flag used by the main thread to tell the copyAsync thread that the main thread is done. */
+    private volatile boolean fileInClosed;
+    /** Flag used by the copyAsync thread to tell the main thread that copyAsync terminated abnormally. */
+    private volatile Throwable throwable;
+
+    FileBackedInputStream(InputStream in, int memLimit, Executor cachedThreadPool) throws IOException {
+        requireNonNull(in, "in");
+        requireNonNull(cachedThreadPool, "cachedThreadPool");
+
+        // Read the first 'memLimit' bytes immediately.
+        byte[] buf = new byte[memLimit];
+        int remaining = memLimit, length = 0, count;
+        while (remaining > 0 && (count = in.read(buf, length, remaining)) != -1) {
+            length += count;
+            remaining -= count;
+        }
+        this.memIn = new ByteArrayInputStream(buf, 0, length);
+
+        if (length < memLimit) {
+            // All content fits in memory
+            in.close();
+            this.sync = null;
+            this.fileIn = null;
+            this.fileInClosed = true;
+        } else {
+            // First 'memLimit' bytes are in memory. Asynchronously download the rest to a file as fast as possible.
+            // The reader can still read from FileBackedInputStream while the download is in progress.
+            this.sync = new Sync();
+            this.file = File.createTempFile("dw-jdbc", ".tmp");
+            this.fileIn = new FileInputStream(file);
+            OutputStream fileOut = new FileOutputStream(file);
+            cachedThreadPool.execute(() -> copyAsync(in, fileOut));
+        }
+    }
+
+    private void copyAsync(InputStream source, OutputStream target) {
+        try (InputStream in = source; OutputStream out = target) {
+            byte[] buf = new byte[4096];
+            int count;
+            while (!fileInClosed && (count = in.read(buf)) != -1) {
+                out.write(buf, 0, count);
+                sync.releaseShared(count);
+            }
+        } catch (Throwable t) {
+            throwable = t;
+        } finally {
+            sync.releaseShared(Long.MAX_VALUE);
+            deleteTempFile();
+        }
+    }
+
+    private void checkAsyncException() throws IOException {
+        if (throwable != null) {
+            throw new IOException(throwable.getMessage(), throwable);
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (memIn != null) {
+            int b = memIn.read();
+            if (b != -1) {
+                return b;
+            }
+            memIn = null;
+        }
+        if (fileIn != null) {
```

Comment:
They're not mutually exclusive, it falls through if `memIn` is at EOF.  I'll try to clarify with a comment.